viam-cartographer_internal_test.go

```diff
@@ -0,0 +1,392 @@
+				makeQuaternionFromGenericMap(inputQuat),
+			)
+			test.That(t, pose, test.ShouldResemble, expectedPose)
+			test.That(t, componentRef, test.ShouldEqual, inputComponentRef)
+		})
+
+		t.Run("empty component reference success", func(t *testing.T) {
+			inputPose = commonv1.Pose{X: 5, Y: 5, Z: 5, OX: 0, OY: 0, OZ: 1, Theta: 0}
+			inputQuat = map[string]interface{}{"real": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+			inputComponentRef = ""
+			pose, componentRef, err := svc.GetPosition(context.Background())
+			test.That(t, err, test.ShouldBeNil)
+			expectedPose := spatialmath.NewPose(
+				r3.Vector{inputPose.X, inputPose.Y, inputPose.Z},
+				makeQuaternionFromGenericMap(inputQuat),
+			)
+			test.That(t, pose, test.ShouldResemble, expectedPose)
+			test.That(t, componentRef, test.ShouldEqual, inputComponentRef)
+		})
+
+		t.Run("invalid quat map keys failure", func(t *testing.T) {
+			inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+			inputQuat = map[string]interface{}{"invalid": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+			inputComponentRef = "componentReference"
+			pose, componentRef, err := svc.GetPosition(context.Background())
+			test.That(t, err, test.ShouldBeError)
+			test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
+			test.That(t, pose, test.ShouldBeNil)
+			test.That(t, componentRef, test.ShouldEqual, "")
+		})
+
+		t.Run("invalid quat map values failure", func(t *testing.T) {
+			inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+			inputQuat = map[string]interface{}{"real": "invalid", "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+			inputComponentRef = "componentReference"
+			pose, componentRef, err := svc.GetPosition(context.Background())
+			test.That(t, err, test.ShouldBeError)
+			test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
+			test.That(t, pose, test.ShouldBeNil)
+			test.That(t, componentRef, test.ShouldEqual, "")
+		})
+
+		t.Run("empty quat map failure", func(t *testing.T) {
+			inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+			inputQuat = map[string]interface{}{}
+			inputComponentRef = "componentReference"
+			pose, componentRef, err := svc.GetPosition(context.Background())
+			test.That(t, err, test.ShouldBeError)
+			test.That(t, err.Error(), test.ShouldContainSubstring, "quaternion given, but invalid format detected")
+			test.That(t, pose, test.ShouldBeNil)
+			test.That(t, componentRef, test.ShouldEqual, "")
+		})
+	})
+
+	// Add invalid client
+	mockSLAMClient.GetPositionFunc = func(
+		ctx context.Context,
+		in *v1.GetPositionRequest,
+		opts ...grpc.CallOption,
+	) (*v1.GetPositionResponse, error) {
+		resp, err := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "not_quat")
+		return resp, err
+	}
+
+	t.Run("invalid client no quat key found in extra failure", func(t *testing.T) {
```

Comment:
[opt] This is what I meant we should do

```suggestion
         t.Run("invalid client no quat key found in extra failure", func(t *testing.T) {
	mockSLAMClient.GetPositionFunc = func(
		ctx context.Context,
		in *v1.GetPositionRequest,
		opts ...grpc.CallOption,
	) (*v1.GetPositionResponse, error) {
		resp, err := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "not_quat")
		return resp, err
	}
```